### PR TITLE
fixes #18253 - container image repository discovery

### PR DESCRIPTION
**Pull request by thomasmckay**

### CODE REVIEW

**Comment by thomasmckay** [comment]
@waldenraines - I need help here. When I discover a registry https://registry.access.redhat.com w/ the /v1/search?q=* choices, then choose "create selected" button, it switches to the yum create repos page (after very briefly showing the docker create repos page). If you could please check this PR and help me get the states, breadcrumbs, and other basic niceties working I can continue.

**Comment by thomasmckay** [comment]
latest version does successful yum and docker search, as well as successful yum repo create. ui for table w/ filter, etc. is still wrong

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/docker-discovery.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<span page-title>{{ 'Docker Repository Discovery' | translate }}</span>
+
+<div bst-alerts success-messages="successMessages" error-messages="errorMessages"></div>
+
+<div class="row">
+  <form ng-submit="discover()" name="discoveryForm" class="col-sm-6">
+    <input type="text"
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/docker-discovery.html]: Putting this into a `<div class="form-group">` will fix the bottom padding of the input.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/yum-discovery-create.html [hunk 2]
```diff
@@ -0,0 +1,157 @@
+<span page-title>{{ 'Create Yum Repositories' | translate }}</span>
+
+<div bst-alerts success-messages="successMessages" error-messages="errorMessages"></div>
+
+<form name="productForm" class="form" ng-disabled="creating()" novalidate role="form">
+  <section class="row">
+    <div class="col-sm-6">
+      <h4 translate>Product Options</h4>
+      <div bst-form-group>
+        <dl class="dl-horizontal dl-horizontal-left">
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/yum-discovery-create.html]: Each of these inputs should be in its own form group and `<dl>, <dd>, and <dt>` should not be used.  See http://www.patternfly.org/pattern-library/widgets/#forms for reference.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/yum-discovery-create.html [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        </dl>
+      </div>
+    </div>
+  </section>
+
+  <section>
+    <h4 translate>Repositories to Create</h4>
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search">
+        <div class="input-group">
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/yum-discovery-create.html]: The class (and the `<div>` itself) is unneeded here since this isn't an input group.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/docker-discovery-create.html [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        </dl>
+      </div>
+    </div>
+  </section>
+
+  <section>
+    <h4 translate>Repositories to Create</h4>
+    <div data-extend-template="layouts/partials/table.html">
+      <div data-block="search">
+        <div class="input-group">
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/docker-discovery-create.html]: The class (and the `<div>` itself) is unneeded here since this isn't an input group.

**Comment by waldenraines** [comment]
Whenever I try to create a docker repo via discovery I get:

"Validation failed: Docker upstream name must be a valid docker name"

**Comment by thomasmckay** [comment]
@waldenraines - mind reviewing ui?
@jlsherrill - mind reviewing server?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create.controller.js [hunk 5]
```diff
@@ -1,92 +1,119 @@
...
-            });
+                unprotected: $scope.createRepoChoices.unprotected,
+                'verify_ssl': $scope.createRepoChoices.verifySsl,
+                'upstream_username': $scope.createRepoChoices.upstreamUsername,
+                'upstream_password': $scope.createRepoChoices.upstreamPassword
+            };
+
+            if (repo.contentType === 'docker') {
+                repoParams.url = $scope.createRepoChoices.repositoryUrl;
+            } else {
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create.controller.js]: Do you need this else statement since `repoParams.url` is initialized to `repo.url` above?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery.controller.js [hunk 6]
```diff
@@ -9,37 +9,64 @@
...
-            $scope.discoveryTable.rows = transformRows(task.output);
+            $scope.table.rows = transformRows(task.output);
+            $scope.table.resource.total = $scope.table.rows.length;
+            $scope.table.resource.subtotal = $scope.table.resource.total;
         };
 
         $scope.setupSelected = function () {
+            var url;
+
+            if (!_.startsWith($scope.discovery.url, 'http')) {
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery.controller.js]: `!_`?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 7]
```diff
@@ -1,116 +1,210 @@
...
+    <div class="col-sm-6">
+      <select class="form-control" ng-disabled="creating()"
+              ng-model="createRepoChoices.existingProductId"
+              ng-options="product.id as product.name for product in products"
+              ng-required="createRepoChoices.newProduct === 'false'"/>
+    </div>
+  </div>
+
+  <div class="form-group" ng-show="createRepoChoices.newProduct === 'true'">
+    <label class="col-sm-2 control-label">Name</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Need to translate this.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 8]
```diff
@@ -1,116 +1,210 @@
...
+             class="form-control"
+             ng-model="createRepoChoices.product.name"
+             ng-disabled="creating()"
+             name="name"
+              ng-required="createRepoChoices.newProduct === 'true'"/>
+    </div>
+  </div>
+
+  <div class="form-group" ng-show="createRepoChoices.newProduct === 'true'">
+    <label class="col-sm-2 control-label">Label</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Need to translate this.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 9]
```diff
@@ -1,116 +1,210 @@
...
 
-    <div ng-hide="createRepoChoices.newProduct === 'true'">
-      <div bst-form-group label="{{ 'Product' | translate }}">
-        <select class="form-control" ng-disabled="creating()"
-          ng-model="createRepoChoices.existingProductId"
-          ng-options="product.id as product.name for product in products">
-        </select>
-      </div>
+  <div class="form-group" ng-show="createRepoChoices.newProduct === 'true' && gpgKeys.length !== 0">
+    <label class="col-sm-2 control-label">GPG Key</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 10]
```diff
@@ -1,116 +1,210 @@
...
-
-      <div bst-form-group label="{{ 'GPG Key' | translate }}">
-        <select class="form-control" ng-model="createRepoChoices.product.gpg_key_id"
-          ng-options="gpg_key.id as gpg_key.name for gpg_key in gpgKeys">
-        </select>
-      </div>
+  <h4 translate>Repository Options</h4>
+
+  <div class="form-group" ng-show="discovery.contentType === 'yum'">
+    <label class="col-sm-2 control-label">Serve via HTTP</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 11]
```diff
@@ -1,116 +1,210 @@
...
+      <input type="checkbox"
+             id="unprotected"
+             ng-model="createRepoChoices.unprotected"
+             ng-disabled="creating()"/>
     </div>
-  </section>
+  </div>
+
+  <div class="form-group" ng-show="discovery.contentType === 'docker'">
+    <label class="col-sm-2 control-label">Registry URL</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 12]
```diff
@@ -1,116 +1,210 @@
...
+             class="form-control"
+             ng-model="createRepoChoices.repositoryUrl"
+             ng-disabled="creating()"/>
+    </div>
+  </div>
 
-  <section>
-    <div bst-form-group label="{{ 'Options' | translate }}">
+  <div class="form-group">
+    <label class="col-sm-2 control-label">Verify SSL</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 13]
```diff
@@ -1,116 +1,210 @@
...
-      ng-disabled="creating()"/>
-      <span translate>Serve via HTTP.</span>
+             id="verify_ssl"
+             ng-model="createRepoChoices.verifySsl"
+             ng-disabled="creating()"/>
+    </div>
+  </div>
+
+  <div class="form-group" ng-show="discovery.contentType === 'docker'">
+    <label class="col-sm-2 control-label" for="registryUsername">Registry Username</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html [hunk 14]
```diff
@@ -1,116 +1,210 @@
...
-        ng-disabled="creating() || repo.created"
-        name="repo_label"
-        type="text"
-        class="form-control"
-        required/>
-      </div>
+  </div>
+
+  <div class="form-group" ng-show="discovery.contentType === 'docker'">
+    <label class="col-sm-2 control-label" for="registryPassword">Registry Password</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery-create.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 15]
```diff
@@ -1,86 +1,135 @@
 <span page-title>{{ 'Repository Discovery' | translate }}</span>
 
-<div>
-  <form ng-submit="discover()" name="discoveryForm" class="col-sm-6">
-    <div class="input-group">
+<div class="row">
+  <form ng-submit="discover()" name="discoveryForm" class="form-horizontal">
+    <div class="form-group">
+      <label class="col-sm-2 control-label">Repository Type</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 16]
```diff
@@ -1,86 +1,135 @@
...
-      <input type="text"
-             ng-pattern="/^(file:\/\/|http:\/\/|https:\/\/).+/"
-             class="form-control input-lg"
-             placeholder="{{ 'URL to discover' | translate }}"
-             ng-model="discovery.url"
-             ng-disabled="discovery.pending"
-             name="discoveryUrl"
-             required/>
+    <div class="form-group" ng-show="discovery.contentType === 'yum'">
+      <label class="col-sm-2 control-label" for="urlToDiscover">URL to Discover</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 17]
```diff
@@ -1,86 +1,135 @@
...
-      <span class="input-group-btn input-group-lg">
-        <button ng-show="!discovery.pending && permitted('edit_products')"
-               ng-disabled="discoveryForm.discoveryUrl.$invalid"
-               translate
-               type="submit"
-               class="btn btn-primary btn-lg">
-          Discover
-        </button>
+    <div class="form-group" ng-show="discovery.contentType === 'docker'">
+      <label class="col-sm-2 control-label" for="registryToDiscover">Registry to Discover</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 18]
```diff
@@ -1,86 +1,135 @@
...
-                ng-disabled="discovery.working"
-                ng-show="discovery.pending && permitted('edit_products')"
-                ng-click="cancelDiscovery()"
-                class="btn btn-primary btn-lg">
-          <i class="fa fa-spinner fa-spin" ng-show="discovery.pending"></i>
-          <span translate>Cancel</span>
-        </button>
-      </span>
+    <div class="form-group" ng-show="discovery.contentType === 'docker'">
+      <label class="col-sm-2 control-label" for="registryUsername">Registry Username</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 19]
```diff
@@ -1,86 +1,135 @@
...
+               type="text"
+               class="form-control"
+               ng-model="discovery.upstreamUsername"
+               ng-disabled="discovery.working"
+               name="upstreamUsername"/>
+      </div>
+    </div>
 
+    <div class="form-group" ng-show="discovery.contentType === 'docker'">
+      <label class="col-sm-2 control-label" for="registryPassword">Registry Password</label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: Missing translation.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html [hunk 20]
```diff
@@ -1,86 +1,135 @@
...
+      </div>
     </div>
-  </form>
-</div>
 
-<section>
-  <div class="row">
-    <h3 class="col-sm-12" translate>Results</h3>
+    <div class="form-group">
+      <label class="col-sm-2 control-label"></label>
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: No label?

**Comment by waldenraines** [comment]
We don't use this type of form anywhere else in the application any longer (or at least shouldn't after https://github.com/Katello/katello/pull/6488):

screen shot 2017-03-02 at 10 13 39 am

Please use the forms shown here:  http://getbootstrap.com/css/#forms

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create.controller.js [hunk 21]
```diff
@@ -1,92 +1,119 @@
...
-            currentlyCreating.form.$invalid = true;
-        }
+angular.module('Bastion.products').controller('DiscoveryCreateController',
+    ['$scope', '$q', 'CurrentOrganization', 'Product', 'Repository', 'GPGKey', 'FormUtils', 'translate',
+    function ($scope, $q, CurrentOrganization, Product, Repository, GPGKey, FormUtils, translate) {
+
+        $scope.errorMessages = [];
+        $scope.successMessages = [];
+
+        $scope.table = {
```
Comment (thomasmckay) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create.controller.js]: Happy to correct if you show me how. I tried for hours to get tables working with the data passed in.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery.controller.js (hunk 6 above)
Comment (thomasmckay) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery.controller.js]: Is this incorrect? What is equivalent of "doesn't start with"?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html (hunk 20 above)
Comment (thomasmckay) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/views/discovery.html]: This was to align the Discover button, should it not be shoved to the right?
repository discovery

File: app/lib/katello/repo_discovery.rb [hunk 22]
```diff
@@ -2,33 +2,57 @@ module Katello
...
+      params[:user] = @upstream_username unless @upstream_username.empty?
+      params[:password] = @upstream_password unless @upstream_password.empty?
+      begin
+        results = RestClient.get(@uri.to_s + "v1/search?q=*", params)
+        JSON.parse(results)['results'].each do |result|
+          @found << result['name']
+        end
+      rescue
+        results = RestClient.get(@uri.to_s + "v2/_catalog", params)
+        @found = JSON.parse(results)['repositories']
```
Comment (jlsherrill) [inline app/lib/katello/repo_discovery.rb]: I'd suggest we sort the output, right nowit appears unsorted

**Comment by waldenraines** [comment]
I am blocked in testing this because when I discover repos I am unable to create the resultant repos.  There is no button or any other UI element to actually create the repos upon discovery.

**Comment by thomasmckay** [comment]
@waldenraines @jlsherrill - updated w/ suggested changes

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create-status.directive.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
+/**
+ * @ngdoc directive
+ * @name Bastion.products.directive:discoveryCreateStatus
+ *
+ * @requires translate
+ *
+ * @description
+ *   Simple directive for encapsulating repository create status.
+ */
+angular.module('Bastion.products').directive('discoveryCreateStatus', ["translate", function (translate) {
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/discovery/discovery-create-status.directive.js]: Meaning that you should delete this file.

**Comment by jlsherrill** [comment]
Found a small issue but it occurs in master too: http://projects.theforeman.org/issues/18865

**Comment by jlsherrill** [comment]
ACK from me, will let @waldenraines give the final ack

**Comment by waldenraines** [comment]
> I am blocked in testing this because when I discover repos I am unable to create the resultant repos. There is no button or any other UI element to actually create the repos upon discovery.

This is still a problem.  It's not clear how you create the repos.

It took me a really long time to figure out I have to click the button I already clicked to create repos.

In this branch:

screen shot 2017-03-13 at 10 14 40 am


Compare that to master, after selecting and clicking "Create Repositories" it is clear how to actually create these repositories:

screen shot 2017-03-13 at 10 09 32 am

We should change the wording of the first button to be "Select Repositories to Create" for both repository types and then move the button on the actual creation page back where it was prior.

**Comment by waldenraines** [comment]
What are the text fields used for in the table below and are they saved when I click the button?

screen shot 2017-03-13 at 10 14 40 am

**Comment by thomasmckay** [comment]
@waldenraines 
+ We could change the "Create Repositories" wording to be "Create Repositories Below". I definitely am not going to put the create button at the bottom of the table since it could very easily be out of sight which would be even more confusing.
+ I could change the column headers to "New Repository Label" and "New Repository Name"

**Comment by waldenraines** [comment]
> We could change the "Create Repositories" wording to be "Create Repositories Below". I definitely am not going to put the create button at the bottom of the table since it could very easily be out of sight which would be even more confusing.

Where the button is now is really confusing because it looks like a table action and because the table doesn't appear to be part of the form.

> I could change the column headers to "New Repository Label" and "New Repository Name"

I don't think that gains us much considering the table is outside of the form.

**Comment by Rohoover** [comment]
@waldenraines @thomasmckay 

I reviewed the screens with Walden earlier today. There are small changes that can be made but they would be more akin to bandaids then really solving the problem. 

It may be better to replace that table with the List View table PatternFly offers, so the URL can be the header and the expansion offer editing. It also allows for actions on the row and actions for the entire table including filtering.

**Comment by waldenraines** [comment]
We've spent a lot of time on this PR to just throw it away at the last minute.

**Comment by thomasmckay** [comment]
@waldenraines @Rohoover - please provide as-detailed code/design as possible so I can correct this weekend. An example html for what the list should look like would be ideal.

**Comment by thomasmckay** [comment]
@waldenraines @Rohoover @jlsherrill - I can't bring myself to make the long form, it is just not usable in my opinion for cases with many repos. I'm going to stand my ground and say accept this as-is or we close it.

**Comment by waldenraines** [comment]
> @waldenraines @Rohoover @jlsherrill - I can't bring myself to make the long form, it is just not usable in my opinion for cases with many repos. I'm going to stand my ground and say accept this as-is or we close it.

If someone else agrees that the UI in this PR is better than what is in master then please feel free to accept these changes.  

The "long form" has apparently been an acceptable solution in master for many years.  I've already said that I would improve the UI after this PR was merged.  And I plan to do that regardless of what gets merged here.

I've been thinking about this more and I think a two step solution (or wizard?) would be better.  First you create/add the repos to the products and then you have the chance to name them (or vice versa).  We could wait to submit everything until the end since there is a constraint with renaming repositories.

**Comment by jlsherrill** [comment]
@waldenraines I actually like what is in this PR better than master, largely due to the formatting in master I think makes it more difficult to visually parse (but this may be personal preference).  I do think i like the idea of a wizard style, so I'd be interested in seeing that.  

tldr; +1 to merging this as is.

**Comment by waldenraines** [comment]
> tldr; +1 to merging this as is.

Okay, then maybe the band-aids @Rohoover [mentioned above](https://github.com/Katello/katello/pull/6564#issuecomment-286815899) should be applied at a minimum?

**Comment by thomasmckay** [comment]
@Rohoover - Can I get a list of suggested wording changes?

**Comment by Rohoover** [comment]
@thomasmckay 

I think I last suggested " Run Creation", or something to that effect. Any thoughts on that?

**Comment by thomasmckay** [comment]
changed to "Run Repository Creation"

**Comment by Rohoover** [comment]
bitmoji

**Comment by jlsherrill** [comment]
ACK from me

**Comment by aleksey-novikov** [comment]
This commit break yum repository creation: repository URL points to discovery URL.

**Comment by thomasmckay** [comment]
created http://projects.theforeman.org/issues/20063
